File: src/bots/refs_utils.php
<?php

namespace WpRefs\Bots\RefsUtils;

/*

Usage:

use function WpRefs\Bots\RefsUtils\str_ends_with;
use function WpRefs\Bots\RefsUtils\str_starts_with;
use function WpRefs\Bots\RefsUtils\rm_str_from_start_and_end;
use function WpRefs\Bots\RefsUtils\remove_start_end_quotes;

*/

if (!function_exists('str_ends_with')) {
    function str_ends_with($string, $endString)
    {
        $len = strlen($endString);
        return substr($string, -$len) === $endString;
    }
}
if (!function_exists('str_starts_with')) {
    function str_starts_with($text, $start)
    {
        return strpos($text, $start) === 0;
    }
}

function rm_str_from_start_and_end(string $text, string $find): string
{
    // ---
    if (!$find) {
        return $text;
    }
    // ---
    $text = trim($text);
    // ---
    if (str_starts_with($text, $find) && str_ends_with($text, $find)) {
        // ---
        // $text = substr($text, strlen($find)); // إزالة $find من البداية
        // $text = substr($text, 0, -strlen($find)); // إزالة $find من النهاية
        // ---
        $text = substr($text, strlen($find), -strlen($find));
    }
    // ---
    return trim($text);
}

function remove_start_end_quotes(string $text): string
{
    // ---
    $text = trim($text);
    // ---
    $text = rm_str_from_start_and_end($text, '"');
    $text = rm_str_from_start_and_end($text, "'");
    // ---
    // echo_test("\n$text\n");
    // ---
    $quote = strpos($text, '"') === false ? '"' : "'";
    // ---
    return $quote . $text . $quote;
}

File: python/src/bots/refs_utils.py
"""
Reference utilities module

Usage:
    from src.bots.refs_utils import (
        str_ends_with, str_starts_with, 
        rm_str_from_start_and_end, remove_start_end_quotes
    )
"""


def str_ends_with(string: str, end_string: str) -> bool:
    """
    Check if string ends with end_string
    
    Args:
        string: String to check
        end_string: Ending to look for
        
    Returns:
        True if string ends with end_string
    """
    return string.endswith(end_string)


def str_starts_with(text: str, start: str) -> bool:
    """
    Check if text starts with start
    
    Args:
        text: Text to check
        start: Starting to look for
        
    Returns:
        True if text starts with start
    """
    return text.startswith(start)


def rm_str_from_start_and_end(text: str, find: str) -> str:
    """
    Remove string from start and end if present on both sides
    
    Args:
        text: Text to process
        find: String to remove
        
    Returns:
        Modified text
    """
    if not find:
        return text
    
    text = text.strip()
    
    if text.startswith(find) and text.endswith(find):
        text = text[len(find):-len(find)]
    
    return text.strip()


def remove_start_end_quotes(text: str) -> str:
    """
    Remove quotes from start and end of text
    
    Args:
        text: Text to process
        
    Returns:
        Text with quotes normalized
    """
    text = text.strip()
    
    text = rm_str_from_start_and_end(text, '"')
    text = rm_str_from_start_and_end(text, "'")
    
    # Choose quote type based on content
    quote = '"' if '"' not in text else "'"
    
    return f"{quote}{text}{quote}"
